<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml"><head><title>R: Phylogenetic Analysis of Identity Classes</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<link rel="stylesheet" type="text/css" href="R.css" />
</head><body>

<table width="100%" summary="page for BuildClusterTree {Seurat}"><tr><td>BuildClusterTree {Seurat}</td><td style="text-align: right;">R Documentation</td></tr></table>

<h2>Phylogenetic Analysis of Identity Classes</h2>

<h3>Description</h3>

<p>Constructs a phylogenetic tree relating the 'average' cell from each
identity class. Tree is estimated based on a distance matrix constructed in
either gene expression space or PCA space.
</p>


<h3>Usage</h3>

<pre>
BuildClusterTree(object, genes.use = NULL, pcs.use = NULL, SNN.use = NULL,
  do.plot = TRUE, do.reorder = FALSE, reorder.numeric = FALSE)
</pre>


<h3>Arguments</h3>

<table summary="R argblock">
<tr valign="top"><td><code>object</code></td>
<td>
<p>Seurat object</p>
</td></tr>
<tr valign="top"><td><code>genes.use</code></td>
<td>
<p>Genes to use for the analysis. Default is the set of
variable genes (object@var.genes). Assumes pcs.use=NULL (tree calculated in
gene expression space)</p>
</td></tr>
<tr valign="top"><td><code>pcs.use</code></td>
<td>
<p>If set, tree is calculated in PCA space, using the
eigenvalue-weighted euclidean distance across these PC scores.</p>
</td></tr>
<tr valign="top"><td><code>SNN.use</code></td>
<td>
<p>If SNN is passed, build tree based on SNN graph connectivity between clusters</p>
</td></tr>
<tr valign="top"><td><code>do.plot</code></td>
<td>
<p>Plot the resulting phylogenetic tree</p>
</td></tr>
<tr valign="top"><td><code>do.reorder</code></td>
<td>
<p>Re-order identity classes (factor ordering), according to
position on the tree. This groups similar classes together which can be
helpful, for example, when drawing violin plots.</p>
</td></tr>
<tr valign="top"><td><code>reorder.numeric</code></td>
<td>
<p>Re-order identity classes according to position on
the tree, assigning a numeric value ('1' is the leftmost node)</p>
</td></tr>
</table>


<h3>Details</h3>

<p>Note that the tree is calculated for an 'average' cell, so gene expression
or PC scores are averaged across all cells in an identity class before the
tree is constructed.
</p>


<h3>Value</h3>

<p>A Seurat object where the cluster tree is stored in
object@cluster.tree[[1]]
</p>

<hr /><div style="text-align: center;">[Package <em>Seurat</em> version 1.3.5 <a href="00Index.html">Index</a>]</div>
</body></html>
