<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml"><head><title>R: Visualize 'features' on a dimensional reduction plot</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<link rel="stylesheet" type="text/css" href="R.css" />
</head><body>

<table width="100%" summary="page for FeaturePlot {Seurat}"><tr><td>FeaturePlot {Seurat}</td><td style="text-align: right;">R Documentation</td></tr></table>

<h2>Visualize 'features' on a dimensional reduction plot</h2>

<h3>Description</h3>

<p>Colors single cells on a dimensional reduction plot according to a 'feature'
(i.e. gene expression, PC scores, number of genes detected, etc.)
</p>


<h3>Usage</h3>

<pre>
FeaturePlot(object, features.plot, dim.1 = 1, dim.2 = 2, cells.use = NULL,
  pt.size = 1, cols.use = c("yellow", "red"), pch.use = 16,
  reduction.use = "tsne", use.imputed = FALSE, nCol = NULL,
  no.axes = FALSE)
</pre>


<h3>Arguments</h3>

<table summary="R argblock">
<tr valign="top"><td><code>object</code></td>
<td>
<p>Seurat object</p>
</td></tr>
<tr valign="top"><td><code>features.plot</code></td>
<td>
<p>Vector of features to plot</p>
</td></tr>
<tr valign="top"><td><code>dim.1</code></td>
<td>
<p>Dimension for x-axis (default 1)</p>
</td></tr>
<tr valign="top"><td><code>dim.2</code></td>
<td>
<p>Dimension for y-axis (default 2)</p>
</td></tr>
<tr valign="top"><td><code>cells.use</code></td>
<td>
<p>Vector of cells to plot (default is all cells)</p>
</td></tr>
<tr valign="top"><td><code>pt.size</code></td>
<td>
<p>Adjust point size for plotting</p>
</td></tr>
<tr valign="top"><td><code>cols.use</code></td>
<td>
<p>The two colors to form the gradient over. Provide as string vector with
the first color corresponding to low values, the second to high. Also accepts a Brewer
color scale or vector of colors. Note: this will bin the data into number of colors provided.</p>
</td></tr>
<tr valign="top"><td><code>pch.use</code></td>
<td>
<p>Pch for plotting</p>
</td></tr>
<tr valign="top"><td><code>reduction.use</code></td>
<td>
<p>Which dimensionality reduction to use. Default is
&quot;tsne&quot;, can also be &quot;pca&quot;, or &quot;ica&quot;, assuming these are precomputed.</p>
</td></tr>
<tr valign="top"><td><code>use.imputed</code></td>
<td>
<p>Use imputed values for gene expression (default is FALSE)</p>
</td></tr>
<tr valign="top"><td><code>nCol</code></td>
<td>
<p>Number of columns to use when plotting multiple features.</p>
</td></tr>
<tr valign="top"><td><code>no.axes</code></td>
<td>
<p>Remove axis labels</p>
</td></tr>
</table>


<h3>Value</h3>

<p>No return value, only a graphical output
</p>

<hr /><div style="text-align: center;">[Package <em>Seurat</em> version 1.3.5 <a href="00Index.html">Index</a>]</div>
</body></html>
