<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml"><head><title>R: Gene expression markers of identity classes defined by a...</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<link rel="stylesheet" type="text/css" href="R.css" />
</head><body>

<table width="100%" summary="page for FindMarkersNode {Seurat}"><tr><td>FindMarkersNode {Seurat}</td><td style="text-align: right;">R Documentation</td></tr></table>

<h2>Gene expression markers of identity classes defined by a phylogenetic clade</h2>

<h3>Description</h3>

<p>Finds markers (differentially expressed genes) based on a branching point (node) in
the phylogenetic tree. Markers that define clusters in the left branch are positive markers.
Markers that define the right branch are negative markers.
</p>


<h3>Usage</h3>

<pre>
FindMarkersNode(object, node, genes.use = NULL, thresh.use = log(2),
  test.use = "bimod", ...)
</pre>


<h3>Arguments</h3>

<table summary="R argblock">
<tr valign="top"><td><code>object</code></td>
<td>
<p>Seurat object</p>
</td></tr>
<tr valign="top"><td><code>node</code></td>
<td>
<p>The node in the phylogenetic tree to use as a branch point</p>
</td></tr>
<tr valign="top"><td><code>genes.use</code></td>
<td>
<p>Genes to test. Default is to use all genes.</p>
</td></tr>
<tr valign="top"><td><code>thresh.use</code></td>
<td>
<p>Limit testing to genes which show, on average, at least
X-fold difference (log-scale) between the two groups of cells. Default is 0.25
Increasing thresh.use speeds up the function, but can miss weaker signals.</p>
</td></tr>
<tr valign="top"><td><code>test.use</code></td>
<td>
<p>Denotes which test to use. Seurat currently implements
&quot;bimod&quot; (likelihood-ratio test for single cell gene expression, McDavid et
al., Bioinformatics, 2011, default), &quot;roc&quot; (standard AUC classifier), &quot;t&quot;
(Students t-test), and &quot;tobit&quot; (Tobit-test for differential gene expression,
as in Trapnell et al., Nature Biotech, 2014)</p>
</td></tr>
<tr valign="top"><td><code>...</code></td>
<td>
<p>Additional arguments passed to FindMarkers</p>
</td></tr>
</table>


<h3>Value</h3>

<p>Matrix containing a ranked list of putative markers, and associated
statistics (p-values, ROC score, etc.)
</p>

<hr /><div style="text-align: center;">[Package <em>Seurat</em> version 1.3.5 <a href="00Index.html">Index</a>]</div>
</body></html>
