<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml"><head><title>R: Gene expression markers for all identity classes</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<link rel="stylesheet" type="text/css" href="R.css" />
</head><body>

<table width="100%" summary="page for FindAllMarkers {Seurat}"><tr><td>FindAllMarkers {Seurat}</td><td style="text-align: right;">R Documentation</td></tr></table>

<h2>Gene expression markers for all identity classes</h2>

<h3>Description</h3>

<p>Finds markers (differentially expressed genes) for each of the identity classes in a dataset
</p>


<h3>Usage</h3>

<pre>
FindAllMarkers(object, ident.1, ident.2 = NULL, genes.use = NULL,
  thresh.use = 0.25, test.use = "bimod", min.pct = 0.1,
  print.bar = TRUE, only.pos = FALSE, max.cells.per.ident = Inf,
  return.thresh = 0.01, do.print = FALSE)
</pre>


<h3>Arguments</h3>

<table summary="R argblock">
<tr valign="top"><td><code>object</code></td>
<td>
<p>Seurat object</p>
</td></tr>
<tr valign="top"><td><code>ident.1</code></td>
<td>
<p>Identity class to define markers for</p>
</td></tr>
<tr valign="top"><td><code>ident.2</code></td>
<td>
<p>A second identity class for comparison. If NULL (default) -
use all other cells for comparison.</p>
</td></tr>
<tr valign="top"><td><code>genes.use</code></td>
<td>
<p>Genes to test. Default is to use all genes.</p>
</td></tr>
<tr valign="top"><td><code>thresh.use</code></td>
<td>
<p>Limit testing to genes which show, on average, at least
X-fold difference (log-scale) between the two groups of cells.
Increasing thresh.use speeds up the function, but can miss weaker signals.</p>
</td></tr>
<tr valign="top"><td><code>test.use</code></td>
<td>
<p>Denotes which test to use. Seurat currently implements
&quot;bimod&quot; (likelihood-ratio test for single cell gene expression, McDavid et
al., Bioinformatics, 2011, default), &quot;roc&quot; (standard AUC classifier), &quot;t&quot;
(Students t-test), and &quot;tobit&quot; (Tobit-test for differential gene expression,
as in Trapnell et al., Nature Biotech, 2014)</p>
</td></tr>
<tr valign="top"><td><code>min.pct</code></td>
<td>
<p>- only test genes that are detected in a minimum fraction of min.pct cells
in either of the two populations. Meant to speed up the function by not testing genes that are very infrequently expression</p>
</td></tr>
<tr valign="top"><td><code>print.bar</code></td>
<td>
<p>Print a progress bar once expression testing begins (uses pbapply to do this)</p>
</td></tr>
<tr valign="top"><td><code>only.pos</code></td>
<td>
<p>Only return positive markers (FALSE by default)</p>
</td></tr>
<tr valign="top"><td><code>max.cells.per.ident</code></td>
<td>
<p>Down sample each identity class to a max number. Default is no downsampling.</p>
</td></tr>
<tr valign="top"><td><code>return.thresh</code></td>
<td>
<p>Only return markers that have a p-value &lt; return.thresh, or a power &gt; return.thresh (if the test is ROC)</p>
</td></tr>
<tr valign="top"><td><code>do.print</code></td>
<td>
<p>FALSE by default. If TRUE, outputs updates on progress.</p>
</td></tr>
</table>


<h3>Value</h3>

<p>Matrix containing a ranked list of putative markers, and associated
statistics (p-values, ROC score, etc.)
</p>

<hr /><div style="text-align: center;">[Package <em>Seurat</em> version 1.3.5 <a href="00Index.html">Index</a>]</div>
</body></html>
