<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml"><head><title>R: ROC-based marker discovery</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<link rel="stylesheet" type="text/css" href="R.css" />
</head><body>

<table width="100%" summary="page for MarkerTest {Seurat}"><tr><td>MarkerTest {Seurat}</td><td style="text-align: right;">R Documentation</td></tr></table>

<h2>ROC-based marker discovery</h2>

<h3>Description</h3>

<p>Identifies 'markers' of gene expression using ROC analysis. For each gene,
evaluates (using AUC) a classifier built on that gene alone, to classify
between two groups of cells.
</p>


<h3>Usage</h3>

<pre>
MarkerTest(object, cells.1, cells.2, genes.use = NULL, thresh.use = log(2),
  print.bar = TRUE)
</pre>


<h3>Arguments</h3>

<table summary="R argblock">
<tr valign="top"><td><code>object</code></td>
<td>
<p>Seurat object</p>
</td></tr>
<tr valign="top"><td><code>cells.1</code></td>
<td>
<p>Group 1 cells</p>
</td></tr>
<tr valign="top"><td><code>cells.2</code></td>
<td>
<p>Group 2 cells</p>
</td></tr>
<tr valign="top"><td><code>genes.use</code></td>
<td>
<p>Genes to test. Default is to use all genes.</p>
</td></tr>
<tr valign="top"><td><code>thresh.use</code></td>
<td>
<p>Limit testing to genes which show, on average, at least
X-fold difference (log-scale) between the two groups of cells. Default is 0.25
Increasing thresh.use speeds up the function, but can miss weaker signals.</p>
</td></tr>
<tr valign="top"><td><code>print.bar</code></td>
<td>
<p>Print a progress bar once expression testing begins (uses pbapply to do this)</p>
</td></tr>
</table>


<h3>Details</h3>

<p>An AUC value of 1 means that expression values for this gene alone can
perfectly classify the two groupings (i.e. Each of the cells in cells.1
exhibit a higher level than each of the cells in cells.2). An AUC value of 0
also means there is perfect classification, but in the other direction. A
value of 0.5 implies that the gene has no predictive power to classify the
two groups.
</p>


<h3>Value</h3>

<p>Returns a 'predictive power' (abs(AUC-0.5)) ranked matrix of
putative differentially expressed genes.
</p>

<hr /><div style="text-align: center;">[Package <em>Seurat</em> version 1.3.5 <a href="00Index.html">Index</a>]</div>
</body></html>
